File: msg/COS_Graph.pl
#!/usr/bin/perl -w
use strict;

use cosmic;

sub ShowGraphLesson {
  my $txt = "";
  $txt .= "# MATH introduce graph structures\n";

  $txt .= ShowLine(Op2("define",
		       "make-graph",
		       ProcMultiple(["nodes", "links"],
				    Op2("pair",
					Ref("nodes"), 
					Ref("links")))));

  $txt .= ShowLine(Op2("define",
		       "test-graph",
		       Op("make-graph",
			  ShowList("1", "2", "3", "4"),
			  ShowList(ShowList("1", "2"),
				   ShowList("2", "3"),
				   ShowList("1", "4")))));
		   
  $txt .= ShowLine(Op2("define",
		       "graph-linked",
		       ProcMultiple(["g", "n1", "n2"],
				    Op1("exists",
					Proc("idx",
					     Op("if",
						Op2("and",
						    Op2(">=", Ref("idx"), 0),
						    Op2("<", Ref("idx"),
							Op1("list-length",
							    Op2("list-ref",
								Ref("g"),
								1)))),
						Op2("list=",
						    Op2("list-ref",
							Op2("list-ref",
							    Ref("g"),
							    1),
							 Ref("idx")),
						    ShowList(Ref("n1"),
							     Ref("n2"))),
						Ref("false")))))));
  $txt .= ShowLine(Op2("=",
		       Op("graph-linked", Ref("test-graph"), "1", "2"),
		       Op("true")));
  $txt .= ShowLine(Op2("=",
		       Op("graph-linked", Ref("test-graph"), "1", "3"),
		       Op("false")));
  $txt .= ShowLine(Op2("=",
		       Op("graph-linked", Ref("test-graph"), "2", "4"),
		       Op("false")));
  $txt .= "# 'if' is used a lot in the next definition in place of and/or\n";
  $txt .= "# this is because I haven't established lazy evaluation forms for and/or\n";
  $txt .= "# so this very inefficient algorithm completely bogs down when combined\n";
  $txt .= "# ... during testing with a dumb implementation for 'exists'.\n";
  $txt .= ShowLine(Op2("define",
		       "graph-linked*",
		       ProcMultiple(["g", "n1", "n2"],
				    Op("if",
				       Op2("=", Ref("n1"), Ref("n2")),
				       Ref("true"),
				       Op("if",
					  Op("graph-linked",
					     Ref("g"),
					     Ref("n1"),
					     Ref("n2")),
					  Ref("true"),
					  Op("exists",
					     Proc("n3",
						  Op("if",
						     Op("graph-linked",
							Ref("g"),
							Ref("n1"),
							Ref("n3")),
						     Op("graph-linked*",
							Ref("g"),
							Ref("n3"),
							Ref("n2")),
						     Ref("false")))))))));
							
  $txt .= ShowLine(Op2("=",
		       Op("graph-linked*", Ref("test-graph"), "1", "2"),
		       Op("true")));
  $txt .= ShowLine(Op2("=",
		       Op("graph-linked*", Ref("test-graph"), "1", "3"),
		       Op("true")));
  $txt .= ShowLine(Op2("=",
		       Op("graph-linked*", Ref("test-graph"), "2", "4"),
		       Op("false")));					    

  return $txt;
};


ShowLesson(ShowGraphLesson());

File: msg/COS_Graph.js
#!/usr/bin/env node

const cos = require('./cosmic');

cos.seed(42);

cos.intro("graph:make");
cos.add(`define graph:make | lambda (x:graph:1 x:graph:2) | pair $x:graph:1 $x:graph:2`);

cos.add(`define demo:graph | graph:make
  (vector 1 2 3 4)
  (vector (pair 1 2) (pair 2 3) (pair 1 4))`);

cos.intro("exists:graph:2");
cos.add(`define exists:graph:2 | lambda (x:graph y:graph:1 z:graph:1) |
  exists | ? n |
    if (or (< $n 0) (>= $n | list-length | list-ref $x:graph 1)) $false |
    list= (list-ref (list-ref $x:graph 1) $n) (pair $y:graph:1 $z:graph:1)`);

cos.add(`exists:graph:2 $demo:graph 1 2`);
cos.add(`not | exists:graph:2 $demo:graph 1 3`);
cos.add(`not | exists:graph:2 $demo:graph 2 4`);
cos.add(`exists:graph:2 $demo:graph 1 4`);

cos.intro("exists:graph:2:list");
cos.add(`define exists:graph:2:list | lambda (x:graph y:graph:1 z:graph:1) |
  if (= $y:graph:1 $z:graph:1) $true |
  if (exists:graph:2 $x:graph $y:graph:1 $z:graph:1) $true |
  exists | ? n:graph:1 |
    if (not | exists:graph:2 $x:graph $y:graph:1 $n:graph:1) $false |
    exists:graph:2:list $x:graph $n:graph:1 $z:graph:1`);

cos.add(`exists:graph:2:list $demo:graph 1 2`);
cos.add(`exists:graph:2:list $demo:graph 1 3`);
cos.add(`not | exists:graph:2:list $demo:graph 2 4`);
